shumai/util/memory.ts

```diff
@@ -0,0 +1,73 @@
+import { Tensor } from '../tensor/tensor'
+export const creationTracker: Map<number, Tensor> = new Map()
+export const altTracker: Record<number, Tensor> = {}
+
+export let is_tidy_run = false
+
+export function tidy<T>(fn: (...args: any[]) => T, args: any | any[] = []): T {
+  if (!Array.isArray(args)) args = [args]
+  is_tidy_run = true
+  const result = fn(...args)
+
+  // lazy mark & sweep w stringifed obj to avoid circ ref
+  const prev_seen = new Set<string>()
+  const parseTidyRet = (result: any) => {
+    const res_type = typeof result
+    if (
+      res_type === 'number' ||
+      res_type === 'bigint' ||
+      res_type === 'symbol' ||
+      res_type === 'string' ||
+      res_type === 'boolean' ||
+      res_type === 'undefined' ||
+      result == null
+    ) {
+      return
+    }
+
+    if (result instanceof Tensor) {
+      creationTracker.delete(result.ptr)
+      return
+    }
+
+    if (result instanceof Array) {
+      // array
+      const length = result.length
+      for (let i = 0; i < length; i++) {
+        if (result[i] instanceof Tensor) {
+          creationTracker.delete(result[i].ptr)
+        } else {
+          parseTidyRet(result[i])
+        }
+      }
+      return
+    }
+
+    if (result instanceof Object) {
+      // Object
+      const stringified = JSON.stringify(result)
```

Comment:
I wonder if we can just use `WeakSet<Object>` for `prev_seen` and avoid stringify calls? (instead of `Set<string>`)

Reply:
TIL about `WeakSet`'s haha, that should work and definitely help w performance!